File: assembla_cli/commands/ticket.py
from __future__ import annotations

import click

from assembla_cli.formatting import print_detail, print_json, print_table


@click.group()
def ticket():
    """Manage tickets."""


@ticket.command("list")
@click.option("-s", "--status", help="Filter by status name")
@click.option("-a", "--assignee", help="Filter by assignee login")
@click.option("-m", "--milestone", help="Filter by milestone ID")
@click.option("-p", "--page", default=1, type=int, help="Page number")
@click.option("-n", "--per-page", default=25, type=int, help="Results per page")
@click.option("--json", "as_json", is_flag=True, help="Output as JSON")
@click.pass_context
def ticket_list(ctx, status, assignee, milestone, page, per_page, as_json):
    """List tickets in the space."""
    client = ctx.obj["client"]
    space = ctx.obj["space"]
    params = {"page": page, "per_page": per_page}
    if status:
        params["ticket_status"] = status
    if assignee:
        params["assigned_to_id"] = assignee
    if milestone:
        params["milestone_id"] = milestone
    data = client.get(f"/spaces/{space}/tickets", **params)
    if as_json:
        print_json(data)
    else:
        print_table(
            data or [],
            columns=["number", "summary", "status_name", "assigned_to_id", "priority"],
            headers=["#", "Summary", "Status", "Assignee", "Priority"],
        )


@ticket.command("show")
@click.argument("number", type=int)
@click.option("--json", "as_json", is_flag=True, help="Output as JSON")
@click.pass_context
def ticket_show(ctx, number, as_json):
    """Show a ticket by number."""
    client = ctx.obj["client"]
    space = ctx.obj["space"]
    data = client.get(f"/spaces/{space}/tickets/{number}")
    if as_json:
        print_json(data)
    else:
        print_detail(data, [
            "number", "summary", "description", "status_name",
            "priority", "assigned_to_id", "milestone_id",
            "created_on", "updated_at",
        ])


@ticket.command("create")
@click.option("-t", "--title", required=True, help="Ticket summary/title")
@click.option("-d", "--description", default="", help="Ticket description")
@click.option("-s", "--status", help="Status name")
@click.option("-p", "--priority", type=int, help="Priority (1=highest)")
@click.option("-a", "--assignee", help="Assignee user ID")
@click.option("-m", "--milestone", help="Milestone ID")
@click.option("--json", "as_json", is_flag=True, help="Output as JSON")
@click.pass_context
def ticket_create(ctx, title, description, status, priority, assignee, milestone, as_json):
    """Create a new ticket."""
    client = ctx.obj["client"]
    space = ctx.obj["space"]
    payload: dict = {"ticket": {"summary": title, "description": description}}
    if status:
        payload["ticket"]["status"] = status
    if priority is not None:
        payload["ticket"]["priority"] = priority
    if assignee:
        payload["ticket"]["assigned_to_id"] = assignee
    if milestone:
        payload["ticket"]["milestone_id"] = milestone
    data = client.post(f"/spaces/{space}/tickets", data=payload)
    if as_json:
        print_json(data)
    else:
        click.echo(f"Created ticket #{data['number']}: {data['summary']}")


@ticket.command("update")
@click.argument("number", type=int)
@click.option("-t", "--title", help="New summary/title")
@click.option("-d", "--description", help="New description")
@click.option("-s", "--status", help="New status name")
@click.option("-p", "--priority", type=int, help="New priority")
@click.option("-a", "--assignee", help="New assignee user ID")
@click.option("-m", "--milestone", help="New milestone ID")
@click.option("--json", "as_json", is_flag=True, help="Output as JSON")
@click.pass_context
def ticket_update(ctx, number, title, description, status, priority, assignee, milestone, as_json):
    """Update a ticket by number."""
    client = ctx.obj["client"]
    space = ctx.obj["space"]
    payload: dict = {"ticket": {}}
    if title is not None:
        payload["ticket"]["summary"] = title
    if description is not None:
        payload["ticket"]["description"] = description
    if status is not None:
        payload["ticket"]["status"] = status
    if priority is not None:
        payload["ticket"]["priority"] = priority
    if assignee is not None:
        payload["ticket"]["assigned_to_id"] = assignee
    if milestone is not None:
        payload["ticket"]["milestone_id"] = milestone
    if not payload["ticket"]:
        click.echo("Nothing to update. Provide at least one option.", err=True)
        raise SystemExit(1)
    data = client.put(f"/spaces/{space}/tickets/{number}", data=payload)
    if as_json:
        print_json(data)
    else:
        click.echo(f"Updated ticket #{number}")


@ticket.command("move")
@click.argument("number", type=int)
@click.argument("status")
@click.pass_context
def ticket_move(ctx, number, status):
    """Move a ticket to a new status.

    Usage: assembla ticket move 123 "Ready for Review"
    """
    client = ctx.obj["client"]
    space = ctx.obj["space"]
    client.put(f"/spaces/{space}/tickets/{number}", data={"ticket": {"status": status}})
    click.echo(f"Ticket #{number} moved to \"{status}\"")

File: cmd/ticket.go
package cmd

import (
	"fmt"
	"os"
	"strconv"

	"github.com/eugene-software/assembla-cli/internal"
	"github.com/spf13/cobra"
)

var ticketCmd = &cobra.Command{
	Use:   "ticket",
	Short: "Manage tickets",
}

var ticketListCmd = &cobra.Command{
	Use:   "list",
	Short: "List tickets",
	RunE: func(cmd *cobra.Command, args []string) error {
		space := requireSpace()

		status, _ := cmd.Flags().GetString("status")
		assignee, _ := cmd.Flags().GetString("assignee")
		milestone, _ := cmd.Flags().GetString("milestone")
		page, _ := cmd.Flags().GetInt("page")
		perPage, _ := cmd.Flags().GetInt("per-page")
		asJSON, _ := cmd.Flags().GetBool("json")

		params := map[string]string{
			"page":     strconv.Itoa(page),
			"per_page": strconv.Itoa(perPage),
		}
		if status != "" {
			params["ticket_status"] = status
		}
		if assignee != "" {
			params["assigned_to_id"] = assignee
		}
		if milestone != "" {
			params["milestone_id"] = milestone
		}

		data, err := Client.Get(fmt.Sprintf("/spaces/%s/tickets", space), params)
		if err != nil {
			return err
		}

		if asJSON {
			internal.PrintJSON(data)
		} else {
			internal.PrintTable(
				internal.ToSlice(data),
				[]string{"number", "summary", "status", "assigned_to_id", "priority"},
				[]string{"#", "Summary", "Status", "Assignee", "Priority"},
			)
		}
		return nil
	},
}

var ticketShowCmd = &cobra.Command{
	Use:   "show <number>",
	Short: "Show ticket details",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		space := requireSpace()
		asJSON, _ := cmd.Flags().GetBool("json")

		number := args[0]
		data, err := Client.Get(fmt.Sprintf("/spaces/%s/tickets/%s", space, number), nil)
		if err != nil {
			return err
		}

		if asJSON {
			internal.PrintJSON(data)
		} else {
			internal.PrintDetail(data, []string{
				"number", "summary", "description", "status",
				"priority", "assigned_to_id", "milestone_id",
				"created_on", "updated_at",
			})
		}
		return nil
	},
}

var ticketCreateCmd = &cobra.Command{
	Use:   "create",
	Short: "Create a new ticket",
	RunE: func(cmd *cobra.Command, args []string) error {
		space := requireSpace()

		title, _ := cmd.Flags().GetString("title")
		description, _ := cmd.Flags().GetString("description")
		status, _ := cmd.Flags().GetString("status")
		priority, _ := cmd.Flags().GetInt("priority")
		assignee, _ := cmd.Flags().GetString("assignee")
		milestone, _ := cmd.Flags().GetString("milestone")
		asJSON, _ := cmd.Flags().GetBool("json")

		ticket := map[string]interface{}{
			"summary":     title,
			"description": description,
		}
		if status != "" {
			ticket["status"] = status
		}
		if cmd.Flags().Changed("priority") {
			ticket["priority"] = priority
		}
		if assignee != "" {
			ticket["assigned_to_id"] = assignee
		}
		if milestone != "" {
			ticket["milestone_id"] = milestone
		}

		payload := map[string]interface{}{
			"ticket": ticket,
		}

		data, err := Client.Post(fmt.Sprintf("/spaces/%s/tickets", space), payload)
		if err != nil {
			return err
		}

		if asJSON {
			internal.PrintJSON(data)
		} else {
			dataMap, ok := data.(map[string]interface{})
			if ok {
				fmt.Printf("Created ticket #%v: %v\n", dataMap["number"], dataMap["summary"])
			}
		}
		return nil
	},
}

var ticketUpdateCmd = &cobra.Command{
	Use:   "update <number>",
	Short: "Update an existing ticket",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		space := requireSpace()
		asJSON, _ := cmd.Flags().GetBool("json")
		number := args[0]

		ticket := map[string]interface{}{}

		if cmd.Flags().Changed("title") {
			v, _ := cmd.Flags().GetString("title")
			ticket["summary"] = v
		}
		if cmd.Flags().Changed("description") {
			v, _ := cmd.Flags().GetString("description")
			ticket["description"] = v
		}
		if cmd.Flags().Changed("status") {
			v, _ := cmd.Flags().GetString("status")
			ticket["status"] = v
		}
		if cmd.Flags().Changed("priority") {
			v, _ := cmd.Flags().GetInt("priority")
			ticket["priority"] = v
		}
		if cmd.Flags().Changed("assignee") {
			v, _ := cmd.Flags().GetString("assignee")
			ticket["assigned_to_id"] = v
		}
		if cmd.Flags().Changed("milestone") {
			v, _ := cmd.Flags().GetString("milestone")
			ticket["milestone_id"] = v
		}

		if len(ticket) == 0 {
			fmt.Fprintln(os.Stderr, "Nothing to update.")
			os.Exit(1)
		}

		payload := map[string]interface{}{
			"ticket": ticket,
		}

		data, err := Client.Put(fmt.Sprintf("/spaces/%s/tickets/%s", space, number), payload)
		if err != nil {
			return err
		}

		if asJSON {
			internal.PrintJSON(data)
		} else {
			fmt.Printf("Updated ticket #%s\n", number)
		}
		return nil
	},
}

var ticketMoveCmd = &cobra.Command{
	Use:   "move <number> <status>",
	Short: "Move a ticket to a new status",
	Args:  cobra.ExactArgs(2),
	RunE: func(cmd *cobra.Command, args []string) error {
		space := requireSpace()
		number := args[0]
		status := args[1]

		payload := map[string]interface{}{
			"ticket": map[string]interface{}{
				"status": status,
			},
		}

		_, err := Client.Put(fmt.Sprintf("/spaces/%s/tickets/%s", space, number), payload)
		if err != nil {
			return err
		}

		fmt.Printf("Ticket #%s moved to \"%s\"\n", number, status)
		return nil
	},
}

func requireSpace() string {
	if Space == "" {
		fmt.Fprintln(os.Stderr, "Error: No space specified. Use --space flag or set in config.")
		os.Exit(1)
	}
	return Space
}

func init() {
	// ticket list flags
	ticketListCmd.Flags().StringP("status", "s", "", "Filter by status")
	ticketListCmd.Flags().StringP("assignee", "a", "", "Filter by assignee ID")
	ticketListCmd.Flags().StringP("milestone", "m", "", "Filter by milestone ID")
	ticketListCmd.Flags().IntP("page", "p", 1, "Page number")
	ticketListCmd.Flags().IntP("per-page", "n", 25, "Results per page")
	ticketListCmd.Flags().Bool("json", false, "Output as JSON")

	// ticket show flags
	ticketShowCmd.Flags().Bool("json", false, "Output as JSON")

	// ticket create flags
	ticketCreateCmd.Flags().StringP("title", "t", "", "Ticket title (required)")
	ticketCreateCmd.MarkFlagRequired("title")
	ticketCreateCmd.Flags().StringP("description", "d", "", "Ticket description")
	ticketCreateCmd.Flags().StringP("status", "s", "", "Ticket status")
	ticketCreateCmd.Flags().IntP("priority", "p", 0, "Ticket priority")
	ticketCreateCmd.Flags().StringP("assignee", "a", "", "Assignee ID")
	ticketCreateCmd.Flags().StringP("milestone", "m", "", "Milestone ID")
	ticketCreateCmd.Flags().Bool("json", false, "Output as JSON")

	// ticket update flags
	ticketUpdateCmd.Flags().StringP("title", "t", "", "Ticket title")
	ticketUpdateCmd.Flags().StringP("description", "d", "", "Ticket description")
	ticketUpdateCmd.Flags().StringP("status", "s", "", "Ticket status")
	ticketUpdateCmd.Flags().IntP("priority", "p", 0, "Ticket priority")
	ticketUpdateCmd.Flags().StringP("assignee", "a", "", "Assignee ID")
	ticketUpdateCmd.Flags().StringP("milestone", "m", "", "Milestone ID")
	ticketUpdateCmd.Flags().Bool("json", false, "Output as JSON")

	ticketCmd.AddCommand(ticketListCmd)
	ticketCmd.AddCommand(ticketShowCmd)
	ticketCmd.AddCommand(ticketCreateCmd)
	ticketCmd.AddCommand(ticketUpdateCmd)
	ticketCmd.AddCommand(ticketMoveCmd)
}
